R/response_process.R

```diff
@@ -0,0 +1,165 @@
+#' Process a Google API response
+#'
+#' @description
+#' `response_process()` is intended primarily for internal use in client
+#' packages that provide high-level wrappers for users. Typically applied as the
+#' final step in this sequence of calls:
+#'   * Request prepared with [request_build()].
+#'   * Request made with [request_make()].
+#'   * Response processed with [response_process()].
+#'
+#' All that's needed for a successful request is to parse the JSON extracted via
+#' `httr::content()`. Therefore, the main point of `response_process()` is to
+#' handle less happy outcomes:
+#'   * Status codes in the 400s (client error) and 500s (server error). The
+#'     structure of the error payload varies across Google APIs and we try to
+#'     create a useful message for all variants we know about.
+#'   * Non-JSON content type, such as HTML.
+#'   * Status code in the 100s (information) or 300s (redirection). These are
```

Comment:
@craigcitro Is my impression correct? That we should not be expecting statii in the 100s or 300s. I've never seen it, but I'd love your input.

Reply:
I've never seen an API return a value in the `1xx` range. I don't think that's written in stone anywhere, but I'd be really surprised.

I think `3xx` errors are possible, eg when fetching data -- BQ may be a case where you encounter this. GCS documents [this possibility](https://cloud.google.com/storage/docs/json_api/v1/status-codes).

There's also [this page](https://cloud.google.com/apis/design/errors#http_mapping) with a semi-authoritative answer about errors and status codes, though it's mostly for "newer" APIs.

Reply:
OK thanks. I have read that page and have built in some of this RPC information. I think we are OK for a first release and if our logic proves to be too simplistic, people will tell us and we will adjust in the face of some concrete examples.